include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -971,7 +971,7 @@ namespace VectorTools
       ReductionControl     control(rhs.size(), 0., 1e-12, false, false);
       SolverCG<LinearAlgebra::distributed::Vector<Number> > cg(control);
       typename PreconditionChebyshev<MatrixType, LocalVectorType>::AdditionalData data;
-      data.matrix_diagonal_inverse = mass_matrix.get_matrix_diagonal_inverse();
+      data.matrix_diagonal_inverse = mass_matrix.get_matrix_diagonal_inverse().get_vector();
```

Comment:
Yes, you would call 
```
data.preconditioner.reset(new DiagonalPreconditioner<LocalVectorType>());
*data.preconditioner = mass_matrix.get_diagonal_inverse();
```

If we had an `std_cxx11::shared_ptr<DiagonalPreconditioner<VectorType> >` in the `Base` class and returned the shared pointer, we could even share the storage between the two. Do you want to try that?

And one more remark: I recommend against using PreconditionChebyshev for solving a single system - at least in the mass matrix context. It needs to estimate eigenvalues by solving an auxiliary linear system. Until you have estimated the eigenvalues of the mass matrix, you have almost solved the system already. My suggestion is to use a Jacobi preconditioner, i.e., simply what you have already. I should have seen that last week already but it somehow slipped through... ;-)